File: 5-tamper-evident-audit-logging/2-csharp/HashChain.cs
using System.Security.Cryptography;
using System.Text;

namespace AuditLogApi;

// Portable hash-chain core. The canonical serialization (fixed field order joined
// with "|", then sha256 hex) MUST be byte-identical across all four languages.
public static class HashChain
{
    // The 64-zero genesis prevHash anchors the first entry — there is no row before index 0.
    public const string GenesisPrevHash = "0000000000000000000000000000000000000000000000000000000000000000";

    public static string ComputeEntryHash(int index, string prevHash, string actor,
        string action, string target, long timestamp)
    {
        var payload = $"{index}|{prevHash}|{actor}|{action}|{target}|{timestamp}";
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(payload));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}

File: 5-tamper-evident-audit-logging/1-java/src/main/java/com/starci/auditapi/HashChain.java
package com.starci.auditapi;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Portable hash-chain core. The canonical serialization (fixed field order joined
 * with "|", then sha256 hex) MUST be byte-identical across all four languages.
 */
public final class HashChain {

    // The 64-zero genesis prevHash anchors the first entry — there is no row before index 0.
    public static final String GENESIS_PREV_HASH = "0".repeat(64);

    private HashChain() {
    }

    public static String computeEntryHash(int index, String prevHash, String actor,
                                          String action, String target, long timestamp) {
        String payload = index + "|" + prevHash + "|" + actor + "|" + action + "|" + target + "|" + timestamp;
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(payload.getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder(hash.length * 2);
            for (byte b : hash) {
                hex.append(Character.forDigit((b >> 4) & 0xF, 16));
                hex.append(Character.forDigit(b & 0xF, 16));
            }
            return hex.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 not available", e);
        }
    }
}
